### PR TITLE
Add first and last page buttons to transaction page

### LINKED ISSUE #3: Difficult to see latest transactions
Now there are a fair few transactions that have taken place, it takes a long while to navigate to the last transaction page.

Would recommend either reversing the order that the transactions are displayed or adding buttons to the PageNavigation component to go to the first and last page.

### PR DESCRIPTION
**Pull request by mustardpower**
Can now navigate to first and last page via two additional buttons. Need some limitation to prevent navigation past the last page?

Closes #3.

### CODE REVIEW

File: src/client/components/PageNavigation.tsx [hunk 1]
```diff
@@ -2,7 +2,9 @@ import { Button, Flex, FlexProps, Text } from '@chakra-ui/react'
 import { queryTypes, useQueryState } from 'next-usequerystate'
 import React from 'react'
 import { FiArrowLeft, FiArrowRight } from 'react-icons/fi'
+import { MdFirstPage, MdLastPage} from 'react-icons/md'
 import { useTransactionPage } from 'src/services/centralized-coin'
+import { getNumberOfPages } from 'src/services/centralized-coin'
```
Comment (franky47) [inline src/client/components/PageNavigation.tsx]: Could you merge the imports please ?

File: src/client/components/PageNavigation.tsx [hunk 2]
```diff
@@ -11,17 +13,26 @@ function useNavigation() {
     defaultValue: 0,
     ...queryTypes.integer
   })
+  
+  const last = React.useCallback(() => getNumberOfPages().then(
+    (value: number) => { setPage(page => value)
```
Comment (franky47) [inline src/client/components/PageNavigation.tsx]: This line breaks when building, probably just `setPage(value)` ?

File: src/client/components/PageNavigation.tsx [hunk 3]
```diff
@@ -11,17 +13,26 @@ function useNavigation() {
...
+  }), [])
 
   const next = React.useCallback(() => setPage(page => page + 1), [])
   const prev = React.useCallback(
     () => setPage(page => Math.max(0, page - 1)),
     []
   )
+
+  const first = React.useCallback(() => setPage(page => 0), [])
```
Comment (franky47) [inline src/client/components/PageNavigation.tsx]: Same here as above.

**Review by franky47 — CHANGES_REQUESTED** [review]

**Review by franky47 — APPROVED** [review]
LGTM!

**Comment by mustardpower** [comment]
There is actually an error with the number of pages calculation - doesn't currently account for any remaining transactions on the last page that isn't completely full.

File: src/services/centralized-coin.ts [hunk 4]
```diff
@@ -104,3 +104,13 @@ export function usePreloadNeighbouringPages(pageNumber: number) {
...
 }
+
+export async function getNumberOfPages() : Promise<number> {
+  const res = await fetch(
+    `https://www.centralized-coin.com/api/stats`
+  )
+  const data = await res.json()
+  const transactionsPerPage = 50;
+  const numberOfTransactions: number = data.count
+  return Math.floor(numberOfTransactions/ transactionsPerPage);
```
Comment (franky47) [inline src/services/centralized-coin.ts]: This should probably be `Math.ceil`, to account for the last page.

Eg, if there were 55 transactions, 55/50 = 1.1, there would be 2 pages.

**Comment by mustardpower** [comment]
I am unsure about the page number calculation. I checked there are 50 transactions on the page but the api endpoint is returning 2772 at this moment in time. However, in the transaction explorer I see only 2 transactions on the last page. Should be 22?

**Review by franky47 — CHANGES_REQUESTED** [review]

**Review by franky47 — APPROVED** [review]
LGTM! Do you want the coins on your Git committer email ? Otherwise, send me your address at contact@francoisbest.com

**Pull request merged**